### PR TITLE
Add per repository package uniqueness constraints

### PR DESCRIPTION
**Pull request by Manisha15**
Add uniqueness constraint for debian repository versions such that there are no duplicate packages in repository.

fixes #6429
https://pulp.plan.io/issues/6429

### CODE REVIEW

File: pulp_deb/app/migrations/0011_auto_20200505_0937.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    operations = [
+        migrations.AlterUniqueTogether(
+            name='installerpackage',
+            unique_together={('relative_path', 'sha256'), ('package', 'version', 'architecture')},
+        ),
+        migrations.AlterUniqueTogether(
+            name='package',
+            unique_together={('relative_path', 'sha256'), ('package', 'version', 'architecture')},
+        ),
+    ]
```
Comment (mdellweg) [inline pulp_deb/app/migrations/0011_auto_20200505_0937.py:21]: This migration is probably outdated (or even unneeded).

Comment (quba42) [reply pulp_deb/app/migrations/0011_auto_20200505_0937.py:21]: I am pretty sure it is needed (since the change would not show up in my test results until I applied the migration), but probably needs to be regenerated.

Comment (mdellweg) [reply pulp_deb/app/migrations/0011_auto_20200505_0937.py]: There is no connection between `unique_together` and `repo_key_field`.
The former provides uniqueness across all content units (on the database level), and the latter provides uniqueness inside a repository version (with the help of `pulpcore.plugin.repo_version_utils.remove_duplicates`).

Comment (Manisha15) [reply pulp_deb/app/migrations/0011_auto_20200505_0937.py]: This was for older implementation (using uniqueness constraint method). With new approach (repo_key_field) this migration is not generated. Fixed it!

Comment (quba42) [reply pulp_deb/app/migrations/0011_auto_20200505_0937.py]: And now the tests are green, yeah!

File: pulp_deb/app/models/content.py [hunk 2]
```diff
@@ -216,6 +216,8 @@ class Package(BasePackage):
 
     SUFFIX = "deb"
 
+    repo_key_fields = ("package", "version", "architecture")
```
Comment (quba42) [inline pulp_deb/app/models/content.py:219]: @mdellweg Do you think this should move to the `BasePackage` class (so it will apply to `InstallerPackage` as well)?

Comment (mdellweg) [reply pulp_deb/app/models/content.py:219]: Good call. I think, yes. (They are insanely similar)

File: pulp_deb/app/models/content.py [hunk 3]
```diff
@@ -198,6 +198,8 @@ def filename(self, component=""):
             "pool", component, prefix, sourcename, "{}.{}".format(self.name, self.SUFFIX)
         )
 
+    repo_key_fields = ("package", "version", "architecture")
```
Comment (quba42) [inline pulp_deb/app/models/content.py:201]: I am thinking we might need to add a checksum to this, in order to achieve what the Debian policy states.

**Comment by quba42** [comment]
I have rebased just the test without the test for now.
The old state is still available on the `deb_repo_uniqueness_test_v1` branch.

File: pulp_deb/app/models/content.py (hunk 3 above)
Comment (quba42) [reply pulp_deb/app/models/content.py:206]: I think this was a nonsense thought...

**Comment by quba42** [comment]
So, having discussed the behaviour with @mdellweg a bit, I have come to the conclusion that this essentially works as designed, though it is not entirely clear to what extent the design accounts for adding several colliding packages in a single command.

The following remains to be done:

- [ ] Test a couple more edge cases and then come to a decision about what sort of test  coverage we need.
- [ ] Mention the effect of this change in the "Feature overview documentation".

This is certainly a significant improvement over current master, which simply creates inconsistent repo versions that will always fail to publish when there are package uniqueness collisions.

**Comment by quba42** [comment]
Due to the importance of this change, as well as the fact that it largely makes use of existing pulpcore features, I will defer the addition of test coverage to a separate issue.

**Review by quba42 — APPROVED** [review]
